File: python-sdk/hsafa/resources/entities.py
from typing import Any, Dict, Optional

from ..http import HttpClient
from ..sse import SSEStream


class EntitiesResource:
    def __init__(self, http: HttpClient):
        self._http = http

    def create(
        self,
        external_id: Optional[str] = None,
        display_name: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._http.post("/api/entities", {
            "type": "human",
            "externalId": external_id,
            "displayName": display_name,
            "metadata": metadata,
        })

    def create_agent(
        self,
        agent_id: str,
        display_name: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._http.post("/api/entities/agent", {
            "agentId": agent_id,
            "displayName": display_name,
            "metadata": metadata,
        })

    def list(
        self,
        type: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Dict[str, Any]:
        return self._http.get("/api/entities", {"type": type, "limit": limit, "offset": offset})

    def get(self, entity_id: str) -> Dict[str, Any]:
        return self._http.get(f"/api/entities/{entity_id}")

    def update(
        self,
        entity_id: str,
        display_name: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._http.patch(f"/api/entities/{entity_id}", {
            "displayName": display_name,
            "metadata": metadata,
        })

    def delete(self, entity_id: str) -> Dict[str, Any]:
        return self._http.delete(f"/api/entities/{entity_id}")

    def subscribe(self, entity_id: str) -> SSEStream:
        url = f"{self._http.base_url}/api/entities/{entity_id}/stream"
        return SSEStream(url, self._http.get_auth_headers(), reconnect=True).start()

File: hsafa-core/old-hsafa/node-sdk/src/resources/entities.ts
import type { HttpClient } from '../http.js';
import type {
  Entity,
  HsafaStream,
  CreateEntityParams,
  CreateAgentEntityParams,
  UpdateEntityParams,
  ListEntitiesParams,
} from '../types.js';
import { SSEStream } from '../sse.js';

export class EntitiesResource {
  constructor(private http: HttpClient) {}

  async create(params: CreateEntityParams): Promise<{ entity: Entity }> {
    return this.http.post('/api/entities', params);
  }

  async createAgent(params: CreateAgentEntityParams): Promise<{ entity: Entity }> {
    return this.http.post('/api/entities/agent', params);
  }

  async list(params?: ListEntitiesParams): Promise<{ entities: Entity[] }> {
    return this.http.get('/api/entities', {
      type: params?.type,
      limit: params?.limit,
      offset: params?.offset,
    });
  }

  async get(entityId: string): Promise<{ entity: Entity }> {
    return this.http.get(`/api/entities/${entityId}`);
  }

  async update(entityId: string, params: UpdateEntityParams): Promise<{ entity: Entity }> {
    return this.http.patch(`/api/entities/${entityId}`, params);
  }

  async delete(entityId: string): Promise<{ success: boolean }> {
    return this.http.delete(`/api/entities/${entityId}`);
  }

  subscribe(entityId: string): HsafaStream {
    const baseUrl = this.http.getBaseUrl();
    const headers = this.http.getAuthHeaders();
    const url = `${baseUrl}/api/entities/${entityId}/stream`;

    return new SSEStream({
      url,
      headers,
      reconnect: true,
    });
  }
}
